httpclient5/src/main/java/org/apache/hc/client5/http/impl/async/AsyncHttpRequestRetryExec.java

```diff
@@ -0,0 +1,177 @@
+import org.apache.hc.core5.http.HttpRequest;
+import org.apache.hc.core5.http.HttpResponse;
+import org.apache.hc.core5.http.nio.AsyncDataConsumer;
+import org.apache.hc.core5.http.nio.AsyncEntityProducer;
+import org.apache.hc.core5.util.Args;
+import org.apache.hc.core5.util.TimeValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Request executor in the asynchronou srequest execution chain that is
+ * responsible for making a decision whether a request than failed due to
+ * an I/O error or received specific response from the target server should
+ * be re-executed.
+ * <p>
+ * Further responsibilities such as communication with the opposite
+ * endpoint is delegated to the next executor in the request execution
+ * chain.
+ * </p>
+ *
+ * @since 5.0
+ */
+@Contract(threading = ThreadingBehavior.STATELESS)
+@Internal
+public final class AsyncHttpRequestRetryExec implements AsyncExecChainHandler {
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    private final HttpRequestRetryStrategy retryStrategy;
+
+    public AsyncHttpRequestRetryExec(final HttpRequestRetryStrategy retryStrategy) {
+        Args.notNull(retryStrategy, "retryStrategy");
+        this.retryStrategy = retryStrategy;
+    }
+
+    private void internalExecute(
+            final int execCount,
+            final HttpRequest request,
+            final AsyncEntityProducer entityProducer,
+            final AsyncExecChain.Scope scope,
+            final AsyncExecChain chain,
+            final AsyncExecCallback asyncExecCallback) throws HttpException, IOException {
+
+        final String exchangeId = scope.exchangeId;
+
+        chain.proceed(RequestCopier.INSTANCE.copy(request), entityProducer, scope, new AsyncExecCallback() {
+
+            @Override
+            public AsyncDataConsumer handleResponse(
+                    final HttpResponse response,
+                    final EntityDetails entityDetails) throws HttpException, IOException {
+                final HttpClientContext clientContext = scope.clientContext;
+                try {
+                    if (entityProducer != null && !entityProducer.isRepeatable()) {
+                        log.debug("{}: cannot retry non-repeatable request", exchangeId);
+                        return asyncExecCallback.handleResponse(response, entityDetails);
+                    }
+                    if (retryStrategy.retryRequest(response, execCount, clientContext)) {
+                        // TODO What to do here?
+                        //response.close();
+                        final TimeValue nextInterval =
+                                retryStrategy.getRetryInterval(response, execCount, clientContext);
+                        if (TimeValue.isPositive(nextInterval)) {
+                            try {
+                                log.debug("{}: wait for {}", exchangeId, nextInterval);
+                                nextInterval.sleep();
```

Comment:
@michael-o We obviously cannot just block the i/o dispatch thread as that would also block all other channels managed by the same selector. There is actually no easy way of implementing an execution delay with an event driven execution handler. We likely need a worker thread and a `ScheduledExecutorService`. For the time being I am going to remove `#getRetryInterval` call from the async processing chain.

Reply:
Indeed, that is an oversight. This would basically mean that the request is immediately retried which does not correspond to the implementation contract of the default impl. Is that really acceptable?

Reply:
@michael-o I corrected async retry exec handling logic. For now the handler will not make any use of `#getRetryInterval`.

Reply:
This change looks good to me, at better than my code ;-)
Shall we document that the retry interval is ignored? I would add it to the class Javadoc.

Reply:
@michael-o Yes, please, do add a blurb about `#getRetryInterval` being a hint that can be ignored by some implementations. In the 5.1 I would like to re-think the whole retry logic for async execution pipeline but that should not delay 5.0 GA.

Reply:
Alright, will do.